py3status/modules/khal_calendar.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+"""
+Displays upcoming khal events.
+
+Configuration parameters:
+    cache_timeout: refresh interval for this module (default 60)
+    config_path: Path to khal configuration file. The default None resolves to /home/$USER/.config/khal/config (default None)
+    date_end: Until which datetime the module searches for events (default 'eod')
+    format: display format for this module (default '{appointments}')
+    output_format: khal conform format for displaying event output (default '{start-time} {title}')
+
+Format placeholders:
+    {appointments} list of events in time range
+
+Requires:
+    khal: https://github.com/pimutils/khal
+
+@author @xenrox
+@license BSD
+
+SAMPLE OUTPUT
+{'full_text': '13:00 Eat lunch'}
+
+"""
+from datetime import datetime
+from re import compile as re_compile
+from khal.settings import get_config
+from khal.cli import build_collection
+from khal.controllers import khal_list
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    cache_timeout = 60
+    config_path = None
+    date_end = "eod"
+    format = "{appointments}"
+    output_format = "{start-time} {title}"
+
+    def _format_output(self, output):
+        ansi_escape = re_compile(r"\x1B\[[0-?]*[ -/]*[@-~]")
+        return ansi_escape.sub("", output)
+
+    def _init_config(self):
+        self.config = get_config(self.config_path)
+        self.collection = build_collection(self.config, None)
+        self.datetimeformat = self.config["locale"]["datetimeformat"]
```

Comment:
Many modules have `post_config_hook()` for initialization. Some of the lines here could go under `post_config_hook`. The other line is `re_compile` (but not `sub`). See if it is possible to get events without colors (or turning colors off before getting events).

Reply:
I already tried that. I can put `config` and `datetimeformat` there, but not `collection` because it is a SQLite object and that would cause threading errors.

Reply:
>I could try to go the same way as... and simply get the direct cli output

`khal --no-color list --day-format "{name}" --format "{start-time}|SEPARATOR|{title}"` does not print the output correctly (imo) making it impossible to match values to the placeholders.
* Expected `16:30|SEPARATOR|Coffee Break`
* Got `|SEPARATOR|16:30 Coffee Break`

`--day-format` is nice too. Can have `Today`, `Tomorrow` (limitation). If you want more flexiblity, you could hardcode `--day-format '{date-long}'` and add strftime support `%a` --> `Thu`. Module `gcal` does this too if you're wondering.

>but i haven't found a way yet to get rid of the ANSI escape code

Not an issue. Just wondering if there is an option to turn off colors. Could ask them.

Reply:
>If you want more flexiblity, you could hardcode --day-format '{date-long}' and add strftime support %a --> Thu. Module gcal does this too if you're wondering

Thats sounds like a great idea. Only problem i see is that the function `khal_list` returns a list with dates and events and i don't know for sure which element is a date and which is an event. I haven't really coded a parses or something similar in python in a long time so maybe you could give me some pointer. Otherwise i have like two ideas: 
1. Bruteforce try to convert unknown element with datetime help. If it doesn't work we have an event (try except way) 
2. I could hardcore a little helper to recognize dates (for example: --day-format "{date-long}HELPERSTRING"

Reply:
>I haven't really coded a parses or something similar in python in a long time so maybe you could give me some pointer.

See `weather_yahoo.py` with `DATETIME_FORECAST`. The module is removed due to full API change by Yahoo/Oath. You can get it via `git checkout 24f03 -- weather_yahoo.py`.

`2)` sounds sensible. `--day-format` looks to be working okay with STRING position.